exonum/src/api/ext.rs

```diff
@@ -0,0 +1,743 @@
+/// just by mentioning a type.
+///
+/// [`ApiBuilder`]: struct.ApiBuilder.html
+/// [`Endpoint`]: trait.Endpoint.html
+pub trait FromContext {
+    /// Creates an instance from the API context.
+    fn from_context(context: &ApiContext) -> Self;
+}
+
+/// Supported subset of HTTP methods.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Method {
+    /// Marker for a safe endpoint, i.e., an endpoint without noticeable side effects.
+    ///
+    /// Named by analogy with HTTP `GET` method.
+    Get,
+
+    /// Marker for an unsafe endpoint, which is mainly called for side effects
+    /// (e.g., pushing a transaction into the memory pool).
+    ///
+    /// Named by analogy with HTTP `POST` method.
+    Post,
+}
+
+/// Definition of an `Endpoint` used to read information from the blockchain.
+///
+/// The macro implements all that is necessary for working with the endpoint
+/// with the exception of the [`Endpoint`] trait itself.
+///
+/// [`Endpoint`]: api/ext/struct.Endpoint.html
+///
+/// # Examples
+///
+/// ```
+/// # #[macro_use] extern crate exonum;
+/// use exonum::api::ext::{ApiResult, Endpoint};
+/// # use exonum::crypto::PublicKey;
+/// # use exonum::storage::Snapshot;
+///
+/// read_request! {
+///     /// Gets the balance of a particular wallet.
+///     @(ID = "balance")
+///     pub GetBalance(PublicKey) -> Option<u64>;
+/// }
+///
+/// // Service schema containing balances.
+/// struct Schema { /* ... */ }
+///
+/// impl Schema {
+/// #  fn new<S: AsRef<Snapshot>>(snapshot: S) -> Self { Schema { } }
+///    pub fn balance(&self, key: &PublicKey) -> Option<u64> {
+///       // ...
+/// #     Some(42)
+///   }
+/// }
+///
+/// impl Endpoint for GetBalance {
+///     fn handle(&self, key: PublicKey) -> ApiResult<Option<u64>> {
+///         let schema = Schema::new(self.as_ref().snapshot());
+///         Ok(schema.balance(&key))
+///     }
+/// }
+/// # fn main() {}
+/// ```
+#[macro_export]
+macro_rules! read_request {
+    // No visibility specifier
+    (
+        $(#[$attr:meta])*
+        @(ID = $id:expr) $name:ident($req:ty) -> $resp:ty;
+    ) => {
+        $(#[$attr])*
+        #[derive(Debug, Clone)]
+        pub struct $name($crate::blockchain::Blockchain);
+
+        read_request!(@implement ($id), $name, $req, $resp);
+    };
+
+    // `pub` visibility specifier
+    (
+        $(#[$attr:meta])*
+        @(ID = $id:expr) pub $name:ident($req:ty) -> $resp:ty;
+    ) => {
+        $(#[$attr])*
+        #[derive(Debug, Clone)]
+        pub struct $name($crate::blockchain::Blockchain);
+
+        read_request!(@implement ($id), $name, $req, $resp);
+    };
+
+    // `pub(..)` visibility specifier
+    // XXX: `pub(in ..)` processing is essentially a hack
+    (
+        $(#[$attr:meta])*
+        @(ID = $id:expr) pub($(in)* $vis:path) $name:ident($req:ty) -> $resp:ty;
+    ) => {
+        $(#[$attr])*
+        #[derive(Debug, Clone)]
+        pub(in $vis) struct $name($crate::blockchain::Blockchain);
```

Comment:
So this creates a bunch of code behind the scene to basically wrap a function. I wonder if we can avoid this code generations and use functions directly instead? I imagine the API to look like

```
        let api = ApiBuilder::new(ctx)
            .add_transactions::<TimeTransactions>()
            .add_get("time", |blockchain: &Blockchain, req: ()| -> ApiResult<SystemTime> { ... });
```

That is, we can use `Fn(&Blochain, U) -> ApiResult<V> where U: Deserialize, V: Serialize` as a `Handler` (or, as an alternative, with `RequestCtx` instead of `Blockchain`, for future extensibility). The main benefit is that there's no macros, so the user doesn't have to decipher what does a CamelCase named structofunction actually do :-) It also allows to close over some over data, and, for example, to gain access to a service-local state from a handler, which might come useful for some services.

Reply:
I'll post a more-or-less detailed memo with design considerations internally approx. tomorrow, so extended discussion will be there :upside_down_face: The use of one (singleton) struct per request in the current design is intentional; it makes it easy to locate all params of the request - it suffices to look at the docs, namely implementation of `EndpointSpec`. And it's on the separate page, too. Additionally, this page may (and imo should) contain at least one example allowing to fix the JSON format for requests / responses of the endpoint for client devs (look at the cryptocurrency demo for an example).